### PR TITLE
Store points into file

### LINKED ISSUE #24: [store points] implementation
demo-implementation

### PR DESCRIPTION
**Pull request by jschleicher**
Closes #24

Todo: More module tests?

### CODE REVIEW

**Comment by Hardthof** [comment]
The test case has 100% coverage for normal use. The edge cases (e.g. empty message, ...) should still be tested.

File: pilz_teleoperation/src/pilz_teleoperation/__init__.py [hunk 1]
```diff
@@ -18,3 +18,4 @@
 from .teleoperation_window import TeleoperationWindow
 from .curses_key_input import CursesKeyInput
 from .terminal_text_window import TerminalTextWindow
+from .ros_message_serializer import RosMessageSerializer
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/__init__.py]:
```suggestion
from .ros_message_serializer import RosMessageSerializer

```

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import genpy
+import rospy
+
+class RosMessageSerializer:
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: allows for better inheritance
```suggestion
class RosMessageSerializer(object):
```

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    Serialization and De-Serialization of ROS messages for convenient file printing
+    """
+    def __init__(self):
+        self._module_imports = {}
+
+    def write_messages_to_file(self, ros_messages, filename):
+        """
+        Writes a file that can be used with 'import' to restore
+        :param ros_messages: iterable of messages to be serialized.
+                tuple contains the variable name used for serialization and a ROS message object
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: why using tuple and not dict since its supposed to be a key, value pair and dict is exactly that.

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        :param filename: target filename, overwritten if existent
+        """
+        with open(filename, 'w') as f:
+            serialized_message = ""
+            for varname, msg in ros_messages:
+                serialized_message += "{} = {}\n".format(varname, self.convert_ros_message_to_python(msg))
+            f.write("\n".join("from {} import {}".format(module, typename) for typename, module in self._module_imports.items()))
+            f.write("\n\n")
+            f.write(serialized_message)
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:43]:
```suggestion
            f.write(serialized_message)
            f.write("\n")
```

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            # end of recursion
+            if isinstance(message, str):
+                return "'{}'".format(message)
+            else:
+                return str(message)
+
+        field_values = []  # ordered list with string-converted attribute values
+        for field_name, field_type in self._get_message_fields(message):
+            val = self.convert_ros_message_to_python(getattr(message, field_name), indentation + 2)
+            field_values.append((field_name,val))
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: again why not using dict?

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        
+        result = ',\n'.join(" "*(indentation+2) + '{} = {}'.format(key,value) for key, value in field_values)
+        if type(message).__name__ not in self._module_imports.keys():
+            self._module_imports[type(message).__name__] = type(message).__module__
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: As i see it, the if is useless, i doubt checking if its in there is much faster, than just putting it in no matter what

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+        return "{}(\n{}\n{})".format(type(message).__name__, result, " " * indentation)
+
+    def _get_message_fields(self, message):
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: can be static

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py (hunk 5 above)
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: maybe extract something in separate methods, and thus get rid of the comments

**Review by DamKoVosh — CHANGES_REQUESTED** [review]
Please also rebase to see if the testsetup still works

**Comment by tobi-v** [comment]
> Please also rebase to see if the testsetup still works

This is already rebased against melodic_devel

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py (hunk 5 above)
Comment (Hardthof) [reply pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: to maintain the order of the values in generated code.
Would you prefer an ordered dict?

Comment (Hardthof) [reply pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: aren't comments a good thing?

Comment (DamKoVosh) [reply pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: Not necessarily => See Clean Code (Robert C. Martin)
Better if you explain yourself in code, and that can easily be done by extracting some code and use the name of a method to explain what you are doing.

Comment (DamKoVosh) [reply pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: not sure if the order is relevant, if it is, ordered dict would be the way to go yes.

Comment (DamKoVosh) [reply pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py]: e.g.
`field_values = get_ordered_string_converted_attributes(message, indentation)`

File: pilz_teleoperation/test/unittests/tst_ros_message_serializer.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import sys
+import rospy
+import pytest
+from pilz_teleoperation.ros_message_serializer import RosMessageSerializer
```
Comment (DamKoVosh) [inline pilz_teleoperation/test/unittests/tst_ros_message_serializer.py]: its in __init__.py so module does not have to be named
```suggestion
from pilz_teleoperation import RosMessageSerializer
```

File: pilz_teleoperation/test/unittests/tst_ros_message_serializer.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
```
Comment (DamKoVosh) [inline pilz_teleoperation/test/unittests/tst_ros_message_serializer.py]: Filename should be test_... so pytest will be found (e.g. pycharm does not recognize them otherwise)
see: [Conventions for Python test discovery](https://docs.pytest.org/en/latest/goodpractices.html)

File: pilz_teleoperation/test/unittests/tst_ros_message_serializer.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
```
Comment (DamKoVosh) [inline pilz_teleoperation/test/unittests/tst_ros_message_serializer.py]: not used
```suggestion
```

File: pilz_teleoperation/test/unittests/tst_ros_message_serializer.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+                                        {"e": None},
+                                        single_pose(),
+                                        multiple_poses()])
+def test_writeback(test_input, tmpdir):
+    """
+    serializes a ros msg, read back and compare with original message
+    """
+
+    # save Pose with unique name
+    module_name = str(tmpdir).split('/')[-1]
```
Comment (DamKoVosh) [inline pilz_teleoperation/test/unittests/tst_ros_message_serializer.py]:
```suggestion
    module_name = tmpdir.basename
```

File: pilz_teleoperation/test/unittests/tst_ros_message_serializer.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    RosMessageSerializer().write_messages_to_file(test_input, str(tmpdir.join(module_name + ".py")))
+
+    # load module
+    sys.path.insert(0, str(tmpdir))
+    readback = __import__(module_name)
+    sys.path.pop(0)
```
Comment (DamKoVosh) [inline pilz_teleoperation/test/unittests/tst_ros_message_serializer.py]: could be replaced with imp.load_source(module_name, file_path)

File: pilz_teleoperation/test/unittests/tst_ros_message_serializer.py (hunk 11 above)
Comment (DamKoVosh) [reply pilz_teleoperation/test/unittests/tst_ros_message_serializer.py]: not sure, why the unique modulname is necessary, since its a unique tmpdir already

test could be reduced to:
```
file_path = str(tmpdir / "test.py")
RosMessageSerializer().write_messages_to_file(test_input, file_path)
read_back = imp.load_source("test", file_path)

for k ...
```

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        :rtype: String
+        """
+        if not isinstance(message, (genpy.Message, rospy.rostime.Time, rospy.rostime.Duration)):
+            # end of recursion
+            return self._message_to_string(message)
+
+        field_values = self._message_to_stringdict(message, indentation)
+        result = ',\n'.join(
+            '{}{} = {}'.format(" " * (indentation + 2), key, value) for key, value in field_values.items())
+        self._module_imports[type(message).__name__] = type(message).__module__
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:58]: type(message).__module__ results in imports like:
```
from geometry_msgs.msg._PoseStamped import PoseStamped
from geometry_msgs.msg._Point import Point
from geometry_msgs.msg._Pose import Pose
from std_msgs.msg._Header import Header
```

so it is using the private modul names, not the finally imported ones. It works, but you may want to think about resolving that, to prevent warnings when using this file

**Review by DamKoVosh — CHANGES_REQUESTED** [review]

File: pilz_teleoperation/test/unittests/tst_ros_message_serializer.py (hunk 12 above)
Comment (DamKoVosh) [reply pilz_teleoperation/test/unittests/tst_ros_message_serializer.py]: just stumbled across [monkeypatching](https://docs.pytest.org/en/latest/monkeypatch.html),
maybe you can add the tmpdir to syspath with that `monkeypatch.syspath_prepend(path)`, gues that depends if you can get access to the tmpdir up there already.

File: pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py (hunk 13 above)
Comment (Hardthof) [reply pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:58]: currently okay as it is.

File: pilz_teleoperation/test/unittests/tst_ros_message_serializer.py (hunk 11 above)
Comment (Hardthof) [reply pilz_teleoperation/test/unittests/tst_ros_message_serializer.py]: imp is deprecated for Python3?

Comment (DamKoVosh) [reply pilz_teleoperation/test/unittests/tst_ros_message_serializer.py]: jes there we would have to use importlib. Don't know if the syntax would stay the same. If not we should find an alternative or leave it as it is.

**Review by DamKoVosh — APPROVED** [review]
